File: lib/wip/command_builder.rb
# frozen_string_literal: true

require 'shellwords'

module Wip
  # Builds the argument arrays for wslc build/exec/run/custom invocations.
  class CommandBuilder
    def initialize(wslc:, config:, environment: Environment.new, dotenv: {})
      @wslc = wslc
      @config = config
      @environment = environment
      @dotenv = dotenv
    end

    def exec(arguments, settings: {}, interactive: true)
      # dependencies: entries don't carry their own name (it's the hash key), so the
      # exec target defaults to @config.container; a commands: entry can still
      # redirect it by setting its own `container:`.
      values = primary_values.merge('container' => required_container).merge(settings)
      command = [@wslc, 'exec']
      command << '-it' if tty?(interactive)
      command.concat(options(values, include_container: true, include_publish: false)).concat(arguments)
    end

    def run(arguments, settings: {}, interactive: true)
      values = primary_values.merge(settings)
      command = [@wslc, 'run']
      command << '--rm' if values['remove']
      command << '-it' if tty?(interactive)
      command.concat(options(values)).push(required(values, 'image')).concat(arguments)
    end

    def up(detach: false)
      values = primary_values
      command = [@wslc, 'run', '--name', required_container]
      command.push('--network', @config.network) if @config.network
      command << '-d' if detach
      command << '-it' if !detach && tty?(true)
      command.concat(options(values)).push(required(values, 'image'))
      command.concat(Shellwords.split(values['command'].to_s)) unless values['command'].to_s.empty?
      command
    end

    def start(detach: false)
      command = [@wslc, 'start', required_container]
      command.push('-a', '-i') unless detach
      command
    end

    def find
      [@wslc, 'list', '--all', '--filter', "name=#{required_container}", '--format', 'json']
    end

    # Mirrors into the volume from a throwaway container. Used for sync.mode:
    # run (compose's default, and mode: container's fallback for when the app
    # container isn't running yet — e.g. just before `up` boots it). The image
    # comes from sync.build's tag (once built via sync_build), else sync.image,
    # else the primary dependencies: entry; compose mode requires one of the
    # first two, since it has no dependencies: entry to fall back to.
    def sync_run
      sync = required_sync
      command = [@wslc, 'run', '--rm']
      sync.volume_specs.each { |spec| command.push('-v', spec) }
      image = sync.build&.fetch('tag') || sync.image || required(primary_values, 'image')
      command.push(image).concat(sync.mirror_command)
    end

    # Builds sync.build's image from a Dockerfile staged in `context` (a caller-
    # managed directory, since the build only reads it once `wslc build` runs).
    # Doesn't touch dependencies: at all, so it works the same under compose
    # mode as it does under container mode.
    def sync_build(context)
      sync = required_sync
      raise ConfigError, 'No sync.build configured in wip.yml' unless sync.build

      [@wslc, 'build', '-t', sync.build['tag'], context]
    end

    # Mirrors from inside the already-running container. Only valid for
    # sync.mode: exec (mode: container's default), since only a container wip
    # itself booted is guaranteed to have both the read-only source mount and
    # the volume attached.
    def sync_exec
      [@wslc, 'exec', required_container, *required_sync.mirror_command]
    end

    def down
      [@wslc, 'stop', required_container]
    end

    # Only reachable under mode: compose-native — mode: compose delegates `wip logs`
    # to the external compose command's own `logs` (ComposeBridge#logs) instead.
    # Single container only, mirroring `wslc`/docker's own `logs`: there's no
    # multi-service log aggregation the way a real compose tool provides.
    def logs(name, follow: true)
      command = [@wslc, 'logs']
      command << '-f' if follow
      command.push(name.to_s)
    end

    def remove
      [@wslc, 'remove', '-f', required_container]
    end

    def network_create
      [@wslc, 'network', 'create', required_network]
    end

    def network_list
      [@wslc, 'network', 'list', '--format', 'json']
    end

    def dependency_up(name, detach: true)
      values = dependency_values(name)
      command = [@wslc, 'run', '--name', name.to_s]
      command.push('--network', @config.network) if @config.network
      command << '-d' if detach
      command.concat(options(values, sync: false)).push(required(values, 'image'))
      command.concat(Shellwords.split(values['command'].to_s)) unless values['command'].to_s.empty?
      command
    end

    def dependency_start(name)
      [@wslc, 'start', name.to_s]
    end

    def dependency_find(name)
      [@wslc, 'list', '--all', '--filter', "name=#{name}", '--format', 'json']
    end

    def dependency_down(name)
      [@wslc, 'stop', name.to_s]
    end

    def dependency_remove(name)
      [@wslc, 'remove', '-f', name.to_s]
    end

    def build(settings:, extra: [])
      values = primary_values.merge(settings)
      context = values['context'] || '.'
      tag = values['tag'] || values['image']
      raise ConfigError, 'Build image/tag must not be empty' if tag.to_s.empty?

      [@wslc, 'build', '-t', tag, *extra, context]
    end

    def custom(name, arguments)
      values = @config.command(name) || raise(ConfigError, "Unknown command: #{name}")
      type = values['type'] || (name.to_s == 'build' ? 'build' : 'exec')
      base = Shellwords.split(values['command'].to_s)
      return build(settings: values, extra: arguments) if type == 'build'

      public_send(type, base + arguments, settings: values, interactive: values.fetch('interactive', false))
    end

    def tty?(requested) = requested && @environment.interactive?

    private

    def options(values, include_container: false, include_publish: true, sync: true)
      result = scalar_options(values)
      merged_env(values).each { |key, value| result.push('-e', "#{key}=#{value}") }
      result.concat(publish_options(values, sync: sync)) if include_publish
      result << required(values, 'container') if include_container
      result
    end

    def scalar_options(values)
      result = []
      result.push('-w', values['workdir']) unless values['workdir'].to_s.empty?
      result.push('-u', values['user']) unless values['user'].to_s.empty?
      result
    end

    def publish_options(values, sync:)
      result = Array(values['ports']).flat_map { |port| ['-p', port.to_s] }
      volume_specs(values, sync: sync).each { |volume| result.push('-v', volume) }
      result
    end

    # With sync configured, a live bind mount of the target (`.:/app`) is
    # swapped for the read-only source mount plus the named volume, so the
    # running app only ever touches the volume.
    def volume_specs(values, sync: true)
      specs = Array(values['volumes']).map(&:to_s)
      settings = sync ? @config.sync : nil
      return specs unless settings

      specs.reject { |spec| settings.replaces?(spec) } + settings.volume_specs
    end

    # .env supplies defaults; env set in wip.yml (primary or per-command) wins on conflict.
    def merged_env(values) = @dotenv.merge(values.fetch('env', {}))

    def required(values, key)
      value = values[key]
      raise ConfigError, "Configured #{key} must not be empty" if value.to_s.empty?

      value
    end

    def required_network
      network = @config.network
      raise ConfigError, 'Configured network must not be empty' if network.to_s.empty?

      network
    end

    def required_sync
      @config.sync || raise(ConfigError, 'No sync: block configured in wip.yml')
    end

    def dependency_values(name)
      @config.dependency(name) || raise(ConfigError, "Unknown dependency: #{name}")
    end

    def required_container
      container = @config.container
      raise ConfigError, 'container: must be set in wip.yml' if container.to_s.empty?

      container
    end

    def primary_values
      container = required_container
      @config.dependency(container) || raise(ConfigError, "No dependencies.#{container} entry " \
                                                          '(check container: in wip.yml)')
    end
  end
end

File: src/Wip.Core/Execution/CommandBuilder.cs
using Wip.Configuration;
using Wip.Platform;
using Wip.Yaml;

namespace Wip.Execution;

/// <summary>
/// Builds the argument arrays for wslc build/exec/run/custom invocations.
/// </summary>
/// <remarks>
/// Everything here returns an argv array that is handed to the process API as a list, never
/// as a command string, so nothing in wip.yml is ever exposed to shell interpretation.
/// </remarks>
public sealed class CommandBuilder
{
    private readonly string wslc;
    private readonly Config config;
    private readonly IEnvironment environment;
    private readonly OrderedDictionary<string, string> dotenv;

    public CommandBuilder(
        string wslc,
        Config config,
        IEnvironment environment,
        OrderedDictionary<string, string>? dotenv = null)
    {
        this.wslc = wslc;
        this.config = config;
        this.environment = environment;
        this.dotenv = dotenv ?? new OrderedDictionary<string, string>(StringComparer.Ordinal);
    }

    public bool Tty(bool requested) => requested && environment.IsInteractive;

    public IReadOnlyList<string> Exec(
        IEnumerable<string> arguments,
        OrderedDictionary<string, object?>? settings = null,
        bool interactive = true)
    {
        // dependencies: entries don't carry their own name (it's the mapping key), so the
        // exec target defaults to config.Container; a commands: entry can still redirect it
        // by setting its own container:.
        var values = PrimaryValues();
        values["container"] = RequiredContainer();
        values = RubyValue.Merge(values, settings);

        var command = new List<string> { wslc, "exec" };
        if (Tty(interactive))
        {
            command.Add("-it");
        }

        command.AddRange(Options(values, includeContainer: true, includePublish: false));
        command.AddRange(arguments);
        return command;
    }

    public IReadOnlyList<string> Run(
        IEnumerable<string> arguments,
        OrderedDictionary<string, object?>? settings = null,
        bool interactive = true)
    {
        var values = RubyValue.Merge(PrimaryValues(), settings);
        var command = new List<string> { wslc, "run" };
        if (RubyValue.IsTruthy(values.GetValueOrDefault("remove")))
        {
            command.Add("--rm");
        }

        if (Tty(interactive))
        {
            command.Add("-it");
        }

        command.AddRange(Options(values));
        command.Add(Required(values, "image"));
        command.AddRange(arguments);
        return command;
    }

    public IReadOnlyList<string> Up(bool detach = false)
    {
        var values = PrimaryValues();
        var command = new List<string> { wslc, "run", "--name", RequiredContainer() };
        if (config.Network is { } network)
        {
            command.Add("--network");
            command.Add(network);
        }

        if (detach)
        {
            command.Add("-d");
        }

        if (!detach && Tty(true))
        {
            command.Add("-it");
        }

        command.AddRange(Options(values));
        command.Add(Required(values, "image"));
        command.AddRange(Shellwords.Split(RubyValue.ToStringValue(values.GetValueOrDefault("command"))));
        return command;
    }

    public IReadOnlyList<string> Start(bool detach = false)
    {
        var command = new List<string> { wslc, "start", RequiredContainer() };
        if (!detach)
        {
            command.Add("-a");
            command.Add("-i");
        }

        return command;
    }

    public IReadOnlyList<string> Find() => Listing(RequiredContainer());

    public IReadOnlyList<string> Stop() => [wslc, "stop", RequiredContainer()];

    public IReadOnlyList<string> Remove() => [wslc, "remove", "-f", RequiredContainer()];

    public IReadOnlyList<string> NetworkCreate() => [wslc, "network", "create", RequiredNetwork()];

    public IReadOnlyList<string> NetworkList() => [wslc, "network", "list", "--format", "json"];

    /// <summary>
    /// Tears down the whole WSLC session, not anything scoped to this <c>wip.yml</c> —
    /// <see cref="Wip.Diagnostics.ErrorInterpreter"/> already points a user at this same
    /// command by hand, as the recovery step for a session-wide mounted-volume limit. Every
    /// other command in this file resolves a specific container or network first; this one
    /// takes no argument at all, because there is nothing project-scoped about it.
    /// </summary>
    public IReadOnlyList<string> SystemSessionTerminate() => [wslc, "system", "session", "terminate"];

    /// <summary>
    /// Mirrors into the volume from a throwaway container, for <c>sync.mode: run</c>. The
    /// image comes from <c>sync.build</c>'s tag, else <c>sync.image</c>, else the primary
    /// dependencies entry; compose mode requires one of the first two, since it has no
    /// dependencies entry to fall back to.
    /// </summary>
    public IReadOnlyList<string> SyncRun()
    {
        var sync = RequiredSync();
        var command = new List<string> { wslc, "run", "--rm" };
        foreach (var spec in sync.VolumeSpecs)
        {
            command.Add("-v");
            command.Add(spec);
        }

        var image = RubyValue.Presence(sync.Build?.GetValueOrDefault("tag"))
                    ?? sync.Image
                    ?? Required(PrimaryValues(), "image");

        command.Add(image);
        command.AddRange(sync.MirrorCommand());
        return command;
    }

    /// <summary>
    /// Builds <c>sync.build</c>'s image from a Dockerfile staged in <paramref name="context"/>,
    /// a caller-managed directory. Doesn't touch dependencies at all, so it works the same
    /// under compose mode as under container mode.
    /// </summary>
    public IReadOnlyList<string> SyncBuild(string context)
    {
        var sync = RequiredSync();
        if (sync.Build is null)
        {
            throw new ConfigException("No sync.build configured in wip.yml");
        }

        return [wslc, "build", "-t", RubyValue.ToStringValue(sync.Build["tag"]), context];
    }

    /// <summary>
    /// Mirrors from inside the already-running container. Only valid for
    /// <c>sync.mode: exec</c>, since only a container wip itself booted is guaranteed to have
    /// both the read-only source mount and the volume attached.
    /// </summary>
    public IReadOnlyList<string> SyncExec()
    {
        var command = new List<string> { wslc, "exec", RequiredContainer() };
        command.AddRange(RequiredSync().MirrorCommand());
        return command;
    }

    /// <summary>
    /// Single container only, mirroring wslc's own <c>logs</c>: there is no multi-service log
    /// aggregation the way a real compose tool provides.
    /// </summary>
    public IReadOnlyList<string> Logs(string name, bool follow = true)
    {
        var command = new List<string> { wslc, "logs" };
        if (follow)
        {
            command.Add("-f");
        }

        command.Add(name);
        return command;
    }

    public IReadOnlyList<string> DependencyUp(string name, bool detach = true)
    {
        var values = DependencyValues(name);
        var command = new List<string> { wslc, "run", "--name", name };
        if (config.Network is { } network)
        {
            command.Add("--network");
            command.Add(network);
        }

        if (detach)
        {
            command.Add("-d");
        }

        command.AddRange(Options(values, sync: false));
        command.Add(Required(values, "image"));
        command.AddRange(Shellwords.Split(RubyValue.ToStringValue(values.GetValueOrDefault("command"))));
        return command;
    }

    public IReadOnlyList<string> DependencyStart(string name) => [wslc, "start", name];

    /// <summary>
    /// Runs a healthcheck's <c>test</c> inside a sidecar. Unlike <see cref="Exec"/>, this
    /// targets <paramref name="name"/> directly rather than <c>config.Container</c>, since a
    /// readiness check runs against whichever dependency it belongs to, not the primary.
    /// </summary>
    public IReadOnlyList<string> DependencyExec(string name, IEnumerable<string> arguments)
    {
        var command = new List<string> { wslc, "exec", name };
        command.AddRange(arguments);
        return command;
    }

    public IReadOnlyList<string> DependencyFind(string name) => Listing(name);

    public IReadOnlyList<string> DependencyStop(string name) => [wslc, "stop", name];

    public IReadOnlyList<string> DependencyRemove(string name) => [wslc, "remove", "-f", name];

    /// <summary>
    /// <c>wslc build</c> reuses matching local layers by default (like <c>docker build</c>
    /// without <c>--pull</c>) — it has no <c>--cache-from</c> flag to ask for that
    /// explicitly, and passing one is a hard error.
    /// </summary>
    public IReadOnlyList<string> Build(OrderedDictionary<string, object?>? settings = null, IEnumerable<string>? extra = null)
    {
        var values = RubyValue.Merge(PrimaryValues(), settings);
        var context = RubyValue.Presence(values.GetValueOrDefault("context")) ?? ".";
        var tag = RubyValue.Presence(values.GetValueOrDefault("tag"))
                  ?? RubyValue.Presence(values.GetValueOrDefault("image"));

        if (tag is null)
        {
            throw new ConfigException("Build image/tag must not be empty");
        }

        var command = new List<string> { wslc, "build", "-t", tag };
        command.AddRange(extra ?? []);
        command.Add(context);
        return command;
    }

    public IReadOnlyList<string> Custom(string name, IEnumerable<string> arguments)
    {
        var values = config.Command(name) ?? throw new ConfigException($"Unknown command: {name}");
        var type = RubyValue.Presence(values.GetValueOrDefault("type")) ?? (name == "build" ? "build" : "exec");
        var argumentList = arguments.ToList();

        if (type == "build")
        {
            return Build(values, argumentList);
        }

        var command = Shellwords.Split(RubyValue.ToStringValue(values.GetValueOrDefault("command"))).ToList();
        command.AddRange(argumentList);

        var interactive = RubyValue.IsTruthy(values.GetValueOrDefault("interactive"));
        return type == "run"
            ? Run(command, values, interactive)
            : Exec(command, values, interactive);
    }

    private IReadOnlyList<string> Listing(string name) =>
        [wslc, "list", "--all", "--filter", $"name={name}", "--format", "json"];

    private List<string> Options(
        OrderedDictionary<string, object?> values,
        bool includeContainer = false,
        bool includePublish = true,
        bool sync = true)
    {
        var result = new List<string>();

        if (RubyValue.Presence(values.GetValueOrDefault("workdir")) is { } workdir)
        {
            result.Add("-w");
            result.Add(workdir);
        }

        if (RubyValue.Presence(values.GetValueOrDefault("user")) is { } user)
        {
            result.Add("-u");
            result.Add(user);
        }

        foreach (var (key, value) in MergedEnvironment(values))
        {
            result.Add("-e");
            result.Add($"{key}={value}");
        }

        if (includePublish)
        {
            foreach (var port in RubyValue.AsArray(values.GetValueOrDefault("ports")))
            {
                result.Add("-p");
                result.Add(RubyValue.ToStringValue(port));
            }

            foreach (var volume in VolumeSpecs(values, sync))
            {
                result.Add("-v");
                result.Add(volume);
            }
        }

        if (includeContainer)
        {
            result.Add(Required(values, "container"));
        }

        return result;
    }

    /// <summary>
    /// With sync configured, a live bind mount of the target (<c>.:/app</c>) is swapped for
    /// the read-only source mount plus the named volume, so the running app only ever touches
    /// the volume.
    /// </summary>
    private List<string> VolumeSpecs(OrderedDictionary<string, object?> values, bool sync)
    {
        var specs = RubyValue.AsArray(values.GetValueOrDefault("volumes")).Select(RubyValue.ToStringValue).ToList();
        var settings = sync ? config.Sync : null;
        if (settings is null)
        {
            return specs;
        }

        var kept = specs.Where(spec => !settings.Replaces(spec)).ToList();
        kept.AddRange(settings.VolumeSpecs);
        return kept;
    }

    /// <summary>.env supplies defaults; env set in wip.yml wins on conflict.</summary>
    private OrderedDictionary<string, string> MergedEnvironment(OrderedDictionary<string, object?> values)
    {
        var merged = new OrderedDictionary<string, string>(StringComparer.Ordinal);
        foreach (var (key, value) in dotenv)
        {
            merged[key] = value;
        }

        if (RubyValue.AsMapping(values.GetValueOrDefault("env")) is { } environment)
        {
            foreach (var (key, value) in environment)
            {
                merged[key] = RubyValue.ToStringValue(value);
            }
        }

        return merged;
    }

    private static string Required(OrderedDictionary<string, object?> values, string key) =>
        RubyValue.Presence(values.GetValueOrDefault(key))
        ?? throw new ConfigException($"Configured {key} must not be empty");

    private string RequiredNetwork() =>
        config.Network ?? throw new ConfigException("Configured network must not be empty");

    private SyncSettings RequiredSync() =>
        config.Sync ?? throw new ConfigException("No sync: block configured in wip.yml");

    private OrderedDictionary<string, object?> DependencyValues(string name) =>
        config.Dependency(name) ?? throw new ConfigException($"Unknown dependency: {name}");

    private string RequiredContainer() =>
        config.Container ?? throw new ConfigException("container: must be set in wip.yml");

    private OrderedDictionary<string, object?> PrimaryValues()
    {
        var container = RequiredContainer();
        return config.Dependency(container)
               ?? throw new ConfigException(
                   $"No dependencies.{container} entry (check container: in wip.yml)");
    }
}
